### PR TITLE
(#1283) Fixed Zip input implementation

### PR DESCRIPTION
**Pull request by fanifieiev**
This PR is for issue #1283

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#1321](https://github.com/yegor256/cactoos/issues/1321) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/C63314D6Z/4450eb39-9c42-4bd0-b28b-a946c274d926); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public void zip() throws Exception {
                 "Can't list files in a directory represented by a path",
                 cnt,
                 // @checkstyle MagicNumber (1 line)
-                Matchers.equalTo(4)
+                Matchers.equalTo(1)
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev please convert this test to `Assertion` and OO matchers instead of static method. Don't forget about changing the assertion message to an assertive sentence ("Must..." instead of "Can't...").

Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev Please replace this matcher with a direct comparison to an iterable of path as strings. (instead of counting them, let's compare them precisely).

**Comment by victornoel** [comment]
@fanifieiev ping?

**Comment by fanifieiev** [comment]
@victornoel Sorry for the delay. Please review again.

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 2]
```diff
@@ -24,51 +24,87 @@
...
+            new File(folder, "A.txt")
+        )
+        ) {
+            out.write("ABC".getBytes());
+        }
+        try (OutputStream out = new OutputStreamTo(
+            new File(folder, "B.txt")
+        )
+        ) {
+            out.write("EFG".getBytes());
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev is it really needed to have all this code to create a file? `new File(...).createNewfile()` is as good, no?

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev or even use `this.temporal.newFile` no?

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 3]
```diff
@@ -24,51 +24,87 @@
...
+        new Assertion<>(
+            "Must zip directory with the same directory structure",
+            new Sticky<>(
+                () -> {
+                    new LengthOf(
+                        new TeeInput(
+                            new InputOf(
+                                new Zip(
+                                    new Directory(folder)
+                                ).stream()
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev `Zip` is already an `Input`, no need to pass it to `InputOf`

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 4]
```diff
@@ -24,51 +24,87 @@
...
+                    ) {
+                        return file.stream().map(ZipEntry::toString).collect(
+                            Collectors.toList()
+                        );
+                    }
+                }
+            ),
+            new MatcherOf<>(
+                sclr -> {
+                    sclr.value().containsAll(
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev directly use `new IsEqual<>(new ListOf<>(..))` here

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 5]
```diff
@@ -24,51 +24,87 @@
...
+                        new ListOf<>(
+                            "abs\\A.txt",
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev is it really `abs`?

File: src/main/java/org/cactoos/io/Zip.java [hunk 6]
```diff
@@ -33,50 +33,58 @@
...
-    public Zip(final Directory origin) {
-        this.origin = origin;
+    public Zip(final Directory dir) {
+        this.directory = dir;
     }
 
     @Override
     public InputStream stream() throws Exception {
         final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final Path parent = new ItemAt<>(0, this.directory).value().getParent();
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev what if it is empty? Please add a test to validate this will work. I believe there is some need to change the interface `Directory` to expose the `Path` itself...

File: src/main/java/org/cactoos/io/Zip.java [hunk 7]
```diff
@@ -33,50 +33,58 @@
...
+            new ForEach<Path>(
+                path -> {
+                    final File file = path.toFile();
+                    final String relative = parent.relativize(path).toString();
+                    if (file.isFile()) {
+                        zip.putNextEntry(new ZipEntry(relative));
+                        try (FileInputStream fis = new FileInputStream(file)) {
+                            zip.write(new BytesOf(new InputOf(fis)).asBytes());
+                        }
+                    } else if (!new Directory(file).iterator().hasNext()) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev what's happening here? it seems this does not go recursively into directories? Can you add a test to validate it works with sub directories?

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 2 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel It is not just creating file, it also puts some content into file

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 3 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Good point

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 4 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel It won't work. `IsEqual` uses equals of List implementation, so it won't check the equality of each element

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 5 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel No, it is a mistake. And also I did missed `return` that's why the test did not fail. Will fix.

File: src/main/java/org/cactoos/io/Zip.java (hunk 6 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel It is never empty. If you pass `Directory `object, it has a path and it has a parent folder:

- For linux: the last parent would be `/`
- For win: the last parent would be `C` or `D`, etc
 Regarding the `I believe there is some need to change the interface Directory to expose the Path itself...`, yes it is required and I will leave a todo.

**Comment by fanifieiev** [comment]
@victornoel Please review again

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 2 above)
Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev I can see that, is this needed? I'm just bothered by the fact that the test contains more things that are not related to what is tested than things that are related to it :)

File: src/main/java/org/cactoos/io/Zip.java [hunk 8]
```diff
@@ -30,53 +30,68 @@
...
+                            new Directory(file).iterator()
+                        );
+                        if (!paths.hasNext()) {
+                            final org.cactoos.text.Joined join = new Joined(
+                                "/", relative, ""
+                            );
+                            zip.putNextEntry(new ZipEntry(join.toString()));
+                        }
                     }
+                    zip.closeEntry();
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev please wrap the code before in a try finally to ensure the entry is always properly closed

File: src/main/java/org/cactoos/io/Zip.java [hunk 9]
```diff
@@ -30,53 +30,68 @@
...
+                        try (FileInputStream fis = new FileInputStream(file)) {
+                            zip.write(new BytesOf(new InputOf(fis)).asBytes());
+                        }
+                    } else {
+                        final Iterator<Path> paths = new Skipped<>(
+                            1,
+                            new Directory(file).iterator()
+                        );
+                        if (!paths.hasNext()) {
+                            final org.cactoos.text.Joined join = new Joined(
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev why the fqdn for Joined? Best to use `Text` on the left side of the assignment anyway

File: src/main/java/org/cactoos/io/Zip.java [hunk 10]
```diff
@@ -30,53 +30,68 @@
...
+                                "/", relative, ""
+                            );
+                            zip.putNextEntry(new ZipEntry(join.toString()));
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev if possible please prefer `asString()` to invoke the behaviour of a `Text`

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev you forgot that

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 11]
```diff
@@ -24,51 +24,87 @@
...
+                    ) {
+                        return new IterableOf<>(
+                            zipped.stream().map(
+                                e -> new TextOf(e.getName())
+                            ).collect(Collectors.toList())
+                        );
+                    }
+                }
+            ),
+            new ScalarHasValue<>(
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev can't we get rid of the `Scalar` stuffs in that test? It seems over-engineered

File: src/main/java/org/cactoos/io/Zip.java (hunk 6 above)
Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev can you add a test where the directory is empty at least?

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 2 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Why not related??? Zipping(aka archiving) is the process to compress content of the files, so the content in files is related to zipping.

File: src/main/java/org/cactoos/io/Zip.java (hunk 6 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel Will do.

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel it is outdated, there is a new code instead of that

File: src/main/java/org/cactoos/io/Zip.java (hunk 9 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel will change

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 11 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel I tried, but this one was the most appropriate in this situation

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 2 above)
Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev I'm sorry I'm not clear: do you test that the content is kept between zipping/unzipping? If yes, then it is related, if not then you can use empty file and the test will still be valid (and simpler to read). Let's keep tests as simple as possible.

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev my comment is still valid, maybe I am not clear? I doubt when reading the code that it does recursively zip the directories, so instead of thinking too much, I'm asking you for a test that validates this expected behaviour.

File: src/main/java/org/cactoos/io/Zip.java (hunk 8 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel it is in `try `block already

Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev let me be clearer: the content of the foreach loop should be in a try, and the `closeEntry` should be in a finally

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 2 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel ok, will use empty files

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel It does not do recursively. The `Directory` object provides the iterable of all the available paths.
The test contains subfolders.
So, the test has the following files and directories:
```
/abc/A.txt # file inside root folder
/abc/B/B.txt # file inside subfolder
/abc/C # empty subdirectory
```

File: src/main/java/org/cactoos/io/Zip.java (hunk 8 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel Why? If the `closeEntry` fails there is no need to iterate over the rest of entries, so 
ZipOutputStream will be closed automatically.
https://docs.oracle.com/javase/7/docs/api/java/util/zip/ZipOutputStream.html#closeEntry()

Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev ok, seems good to me then

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev I missed that, ok. I think it would make sense to have multiple tests for the various aspects being tested and not everything in the same test though:
- empty dir
- multiple nested directories
- nominal behaviour with the correct base path being zipped

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 11 above)
Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev ok, let's keep it like that, I don't see any way to make the test clearer except moving the zipping part out of the assertion argument.

File: src/main/java/org/cactoos/io/Zip.java (hunk 7 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel Agree, but I think we should have at least one test which would have complex case, when there is an empty folder, file in subfolder and just a file inside main folder.
I will also create a separate test for empty dir.

**Comment by fanifieiev** [comment]
@victornoel Please review again.
I still do not like the way of creating directories and files.
Maybe we need to create some EO functionality to create files/dirs elegantly? WDYT?

**Comment by victornoel** [comment]
@fanifieiev 
> I still do not like the way of creating directories and files.
> Maybe we need to create some EO functionality to create files/dirs elegantly? WDYT?

Maybe create an issue to discuss it?

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 12]
```diff
@@ -24,51 +24,113 @@
...
+                            ).asBytes()
+                        );
+                    }
+                    try (ZipFile zipped = new ZipFile(
+                        new File(folder.getParentFile(), "empty.zip")
+                    )
+                    ) {
+                        return zipped.stream().map(
+                            ZipEntry::toString
+                        ).findFirst().get();
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev can we get rid of `stream()` usage? Also this is not good, if there is a bug and there is two files in the zip, only the first one will be returned and the test will pass incorrectly

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 13]
```diff
@@ -24,51 +24,113 @@
...
+                        );
+                    }
+                    try (ZipFile zipped = new ZipFile(
+                        new File(folder.getParentFile(), zipname)
+                    )
+                    ) {
+                        return new IterableOf<>(
+                            zipped.stream().map(
+                                e -> new TextOf(e.getName())
+                            ).collect(Collectors.toList())
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev can we get rid of stream and collect here?

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 12 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel OK, makes sense

**Comment by fanifieiev** [comment]
> @fanifieiev
> 
> > I still do not like the way of creating directories and files.
> > Maybe we need to create some EO functionality to create files/dirs elegantly? WDYT?
> 
> Maybe create an issue to discuss it?

@victornoel Will do

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 13 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel I am not sure how to do without `Stream`, I actually tried, but could not make it work.
If you look inside `stream()` method of `ZipEntry`, you will see a lot of things happening there.

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 12 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Makes sense

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 13 above)
Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev can't you wrap a stream into an `IterableOf` and then use `Mapped` on it for example? I think there should be something like that possible.

**Comment by fanifieiev** [comment]
> > @fanifieiev
> > > I still do not like the way of creating directories and files.
> > > Maybe we need to create some EO functionality to create files/dirs elegantly? WDYT?
> > 
> > 
> > Maybe create an issue to discuss it?
> 
> @victornoel Will do

@victornoel I have created a ticket: https://github.com/yegor256/cactoos/issues/1333

**Comment by fanifieiev** [comment]
@victornoel Please review again

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 14]
```diff
@@ -75,14 +76,21 @@ public void mustZipEmptyDirectory() throws Exception {
                         new File(folder.getParentFile(), "empty.zip")
                     )
                     ) {
-                        return zipped.stream().map(
-                            ZipEntry::toString
-                        ).findFirst().get();
+                        return new IterableOf<>(
+                            new ListOf<>(
+                                new Mapped<>(
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev please use directly the iterable abstraction, you don't need list:
```
return new Mapped<>(
  ZipEntry::getName,
  zipped.stream()::iterator
)
```

should work

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 15]
```diff
@@ -117,9 +125,12 @@ public void mustZipDirectory() throws Exception {
                     )
                     ) {
                         return new IterableOf<>(
-                            zipped.stream().map(
-                                e -> new TextOf(e.getName())
-                            ).collect(Collectors.toList())
+                            new ListOf<>(
+                                new Mapped<>(
+                                    entry -> new TextOf(entry.getName()),
+                                    zipped.stream().iterator()
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev same thing

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 14 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel I need ListOf, because I need to collect all the values into collection before I leave the try block, otherwise it fails to get the values.
I would not do like that, if I could use just as you posted.

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev in that case please remove the wrapping `IterableOf`, it is not needed (a `List` is an `Iterable`).

Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Right, but `IterableOf `has equals that can check the equality of each element, the `ListOf `does not have.

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev have you tried? `ListOf` extends `ListEnvelope` which delegate the `equals` method so it should work as expected IMHO or we should change the matcher so that it works.

Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Have you see the implementation of `equals` of `ListOf ` ???
The implementation of `equals `for `ListOf` is here https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/iterable/IterableEnvelope.java

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev yes, it delegates to the wrapped list, that's what I said

Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel The warpped list in this case is `LinkedList`. It does not check the equality of each element.

Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Ok, will try to change the matcher

**Comment by fanifieiev** [comment]
@victornoel Please review the latest changes

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 14 above)
Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: > The warpped list in this case is LinkedList. It does not check the equality of each element.

@fanifieiev yes it does… that's part of the contract of equals on all `List`… if you don't want to test for order, use a `SetOf` instead

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 16]
```diff
@@ -24,51 +24,121 @@
...
-    public final TemporaryFolder folder = new TemporaryFolder();
+    public final TemporaryFolder temporal = new TemporaryFolder();
+
+    @Test
+    public void mustZipEmptyDirectory() throws Exception {
+        final File folder = this.temporal.newFolder("empty");
+        new Assertion<>(
+            "Must zip empty directory",
+            new Sticky<>(
+                new IterableOf<>(
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev why do you add even more decorators? I propose the following:
- returns a `SetOf` inside the scalar, so that we don't bother about order
- use the matcher `IsEqual` with a `SetOf` there too

And it should work perfectly.

We don't want to use `HasValues` because it is not strict about the content, it only check that some things are present (like `contains()`)

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 14 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel You are right, AbstractList has equals implementation. I though it did not, because it did not work for me.

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 16 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel I need to wrap scalar lambda into iterable, that is why I am doing:
```
new IterableOf<>(
   () -> {...}
)
```

and I use `Sticky `because I need cached value.

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev I'm sorry, but that's way too complex. We should be able to simplify much more this code... what's the point of creating a list of a mapped iterator if after that you return the list's iterator and then you wrap it into an iterable that you make sticky? That's way too complex.

We spent already too much time here, please simplify that with as less as possible classes involved. Take the time to understand what's the expected behaviour of those classes and remove layers instead of adding more.

I see two simple solutions:
- the one you had before based on ScalarHasValue with a Sticky scalar responsible of holding the value and the lambda of the scalar just returning a `ListOf` or a `SetOf` (whichever, it's not important)
- move all this code in the lambda OUT of the assertion and simply store the resulting list or iterable or set or whatever in a variable.

Comment (victornoel) [reply src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev I have created #1334 to explain what is the problem with `iterable.Sticky`.

Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel OK

**Comment by fanifieiev** [comment]
@victornoel Please take a look

File: src/main/java/org/cactoos/io/Zip.java [hunk 17]
```diff
@@ -30,53 +30,69 @@
...
+                            1,
+                            new Directory(file).iterator()
+                        );
+                        if (!paths.hasNext()) {
+                            zip.putNextEntry(
+                                new ZipEntry(
+                                    new Joined("/", relative, "").asString()
+                                )
+                            );
+                        }
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev one last thing: what is the purpose of this introduced else branch here? it wasn't there before and it doesn't seem to solve the problem from the original ticket you are solving.

Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: victornoel this else branch adds empty folder to zip file, if empty folder or subfolder is present.

Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev it seems also complex for what it does, as a REV I'm not happy with it and I don't want to spend more time explaining what bothers me and what should be done.
So I think it would be safer to remove this and add a todo for this and let the work continue in another ticket/PR.

**Comment by fanifieiev** [comment]
@victornoel I removed the functionality of zipping empty folders.Please take a look.

File: src/main/java/org/cactoos/io/Zip.java [hunk 18]
```diff
@@ -33,50 +33,54 @@
...
+                    final String relative = this.directory.path().getParent()
+                        .relativize(path).toString();
+                    final File file = path.toFile();
+                    if (file.isFile()) {
+                        zip.putNextEntry(new ZipEntry(relative));
+                        try (FileInputStream fis = new FileInputStream(file)) {
+                            zip.write(new BytesOf(new InputOf(fis)).asBytes());
+                        }
                     }
+                    zip.closeEntry();
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev this should be inside the `if` : the entry is created only if it is a file.

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 19]
```diff
@@ -24,51 +24,80 @@
...
-        }
+    public void mustZipDirectory() throws Exception {
+        final String zipname = "abc.zip";
+        final File folder = this.temporal.newFolder("abc");
+        new File(folder, "A.txt").createNewFile();
+        new File(folder, "B").mkdir();
+        new File(
+            new File(
+                new Joined(File.separator, folder.getPath(), "B").asString()
+            ), "B.txt"
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev this seems very complex, what is wrong with: `new File(new File(folder, "B"), "B.txt"))` ? You can even allocate `new File(folder, "B")` to a variable to be able to not repeat yourself

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 20]
```diff
@@ -24,51 +24,80 @@
...
+import org.llorllale.cactoos.matchers.ScalarHasValue;
 
 /**
  * Test case for {@link Zip}.
  *
  * @since 0.29
  * @checkstyle JavadocMethodCheck (500 lines)
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
```
Comment (victornoel) [inline src/test/java/org/cactoos/io/ZipTest.java]: @fanifieiev please remove this, it seems the only repeated literal is `abc`, `B.txt` and `A.txt`, so you should be able to allocate the `File` to a variable and refer to their name in the rest of the test.

File: src/main/java/org/cactoos/io/Zip.java (hunk 18 above)
Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel Agree

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 19 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Agree

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 20 above)
Comment (fanifieiev) [reply src/test/java/org/cactoos/io/ZipTest.java]: @victornoel Ok

**Comment by fanifieiev** [comment]
@victornoel Please review

File: src/main/java/org/cactoos/io/Zip.java [hunk 21]
```diff
@@ -33,50 +33,54 @@
...
     public InputStream stream() throws Exception {
         final ByteArrayOutputStream out = new ByteArrayOutputStream();
         try (ZipOutputStream zip = new ZipOutputStream(out)) {
-            for (final Path path : this.origin) {
-                final File file = path.toFile();
-                final ZipEntry entry = new ZipEntry(
-                    file.getPath()
-                );
-                zip.putNextEntry(entry);
-                if (file.isFile()) {
```
Comment (victornoel) [inline src/main/java/org/cactoos/io/Zip.java]: @fanifieiev actually, now that I read this change in its greater context, I realise that the code before your changes was perfectly ok with the `closeEntry()` were it was because it did the `putNextEntry` entry BEFORE the `if`.

I think we should revert to the original behaviour and put back the test for empty directories, everything should be working without problem and there is no need for the `@todo`.

Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel Ok, will revert 2 latest commits then

Comment (victornoel) [reply src/main/java/org/cactoos/io/Zip.java]: @fanifieiev that's not what I said, I'm sorry I wasn't clear:
- you don't need the else branch
- you just need to put back the `putNextEntry` before the `if`
- and keep the `closeEntry` out of the `if` in that case.

In other words: you should revert back to the original code before any of your changes.

And from there:
- only add the relativisation of the path as required by the issue you are solving.
- if you want you can add the foreach for the loop, it makes sense
- then you can use the new tests you made and leave the new one for empty directories.
- if the test for empty directories does not pass, then you can add a todo

But please, do not try to tackle more, looking back, I think I should have refused the PR instead of taking so much of our time :( Sorry about that.

Comment (fanifieiev) [reply src/main/java/org/cactoos/io/Zip.java]: @victornoel if I remove the else branch, the test for empty dir won't pass.
Actually, please refuse this PR, I will create another one later.
Thanks.

**Comment by victornoel** [comment]
@paulodamaso let's refuse this, @fanifieiev will make another PR.
@fanifieiev again, it's too bad we took too long to realise this was going to eat too much of our energy, I'm not very clear as to what is the best strategy in such case, because I thought we could succeed ^^ My main feedback would be to change as less things as possible (relativise the path) while cleaning up stuffs in a purely syntactic way (using `ForEach` for example or `Assertion` in tests) and adding a test to validate the expected behaviour.

**Comment by fanifieiev** [comment]
@paulodamaso I am closing this PR as it has already taken much time.
Will create another PR.

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Code review was too long (15 days), architects (@paulodamaso) were penalized, see [§55](http://www.zerocracy.com/policy.html#55)

**Comment by 0crat** [comment]
Pull request [#1321](https://github.com/yegor256/cactoos/issues/1321) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by sereshqua** [comment]
@victornoel @fanifieiev this PR has the most amount of revisions i've seen in ZC :)

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by fanifieiev** [comment]
> @victornoel @fanifieiev this PR has the most amount of revisions i've seen in ZC :)
@sereshqua So, @victornoel  is the winner and deserves double points :)

**Comment by victornoel** [comment]
@0crat status

**Comment by victornoel** [comment]
@fanifieiev I think it shows more a failure on my part than anything else :sweat_smile:

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/cactoos/issues/1321#issuecomment-605604881))

@victornoel This is what I know about this job in [C63314D6Z](https://www.0crat.com/p/C63314D6Z), as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#1321](https://github.com/yegor256/cactoos/issues/1321) is not in scope
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos%231321%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by sereshqua** [comment]
@0crat quality good

**Comment by 0crat** [comment]
> @0crat quality good ([here](https://github.com/yegor256/cactoos/issues/1321#issuecomment-605666585))

@sereshqua There is no quality review for [#1321](https://github.com/yegor256/cactoos/issues/1321), no performer
